nameservers: Use nmcli to update the nameserver and search info

praveenkumar opened the pull request:
From 4.15, because of ovn-kubernetes the info around nameserver and
search option in the `/etc/resolv.conf` not match with user expectation.
`ovs-configuration` service is run on this node which is a dependency
for kubelet and this make changes in the networking by creating bridge
network and restarting the NetworkManager. As soon as NM restart happen,
changes which are done as part of crc vm post start are vanished.

This PR is going to make sure that changes are done using `nmcli` and
shouldn't removed when NM restart happen. Following steps are done in
this PR
- Start the `ovs-configuration` service, even it is enabled it doesn't
  autostart because it only required by `kubelet-dependencies.target`.
- `ovs-configuration` service always create the network named as
  `ovs-if-br-ex` so we run the nmcli command to update this connection
  and add the nameserver and search option
- Restart the NM to update the /etc/resolv.conf

**Fixes:** Issue #4144

praveenkumar committed:
nameserver: rename CreateResolvFileOnInstance

It is renamed to UpdateResolvFileOnInstance. The resolv.conf file
always present in the VM, this function just update it with user
provided nameserver and our custom dnsmasq instance IP.

praveenkumar committed:
nameservers: Use nmcli to update the nameserver and search info

From 4.15, because of ovn-kubernetes the info around nameserver and
search option in the `/etc/resolv.conf` not match with user expectation.
`ovs-configuration` service is run on this node which is a dependency
for kubelet and this make changes in the networking by creating bridge
network and restarting the NetworkManager. As soon as NM restart happen,
changes which are done as part of crc vm post start are vanished.

This PR is going to make sure that changes are done using `nmcli` and
shouldn't removed when NM restart happen. Following steps are done in
this PR
- Start the `ovs-configuration` service, even it is enabled it doesn't
  autostart because it only required by `kubelet-dependencies.target`.
- `ovs-configuration` service always create the network named as
  `ovs-if-br-ex` so we run the nmcli command to update this connection
  and add the nameserver and search option
- Restart the NM to update the /etc/resolv.conf

cfergeau commented on pkg/crc/network/nameservers.go:
```diff
@@ -31,7 +31,7 @@ func GetResolvValuesFromInstance(sshRunner *ssh.Runner) (*ResolvFileValues, erro
 }
 
-func CreateResolvFileOnInstance(sshRunner *ssh.Runner, resolvFileValues ResolvFileValues) error {
+func UpdateResolvFileOnInstance(sshRunner *ssh.Runner, resolvFileValues ResolvFileValues) error {
```
> resolve.conf file always present in the VM, this function just update it with user provided nameserver and our custom dnsmasq instance IP.

Typo, `resolve.conf` -> `resolv.conf`

This function overwrites the existing `/etc/resolv.conf` with a new file generated from `resolvFileValues`, I don't think Update is better/worse than Create. if you really want to clarify it's name, I'd name it `ReplaceResolvFileOnInstance` (or even `ReplaceInstanceResolvConf`

praveenkumar replied:
But if you check the next commit where we use the `nmcli` then `update` make more sense then replace.

cfergeau replied:
This commit would make more sense if it's moved *after* the next commit then :)

praveenkumar replied:
done.

cfergeau commented on pkg/crc/network/nameservers.go:
```diff
@@ -34,11 +36,35 @@ func GetResolvValuesFromInstance(sshRunner *ssh.Runner) (*ResolvFileValues, erro
-		return fmt.Errorf("Error creating /etc/resolv on instance: %s", err.Error())
+	sd := systemd.NewInstanceSystemdCommander(sshRunner)
+	// Check if ovs-configuration.service exist and if not then it is old bundle and use the same way to
+	// update resolve.conf file
```
Please split all the new code in several functions.

cfergeau commented on pkg/crc/network/nameservers.go:
```diff
@@ -34,11 +36,35 @@ func GetResolvValuesFromInstance(sshRunner *ssh.Runner) (*ResolvFileValues, erro
+	nameservers := resolvFileValues.NameServers[0].IPAddress
+	for _, n := range resolvFileValues.NameServers[1:] {
+		nameservers = fmt.Sprintf("%s,%s", nameservers, n.IPAddress)
+	}
```
https://pkg.go.dev/strings#Join ?

praveenkumar replied:
To use `strings.join` I have to create another slice to capture value `resolvFileValues.NameServer` otherwise I can't .directly use because of type mismatch,

```
strings.Join(resolvFileValues.NameServers,  "," )
```

cfergeau replied:
Hmm, the structure of the ResolvFileValues type might be a bit too complicated, maybe it could be changed to only contain strings instead of struct containing strings
Or you could add this API:
```
@@ -34,6 +35,7 @@ func GetResolvValuesFromInstance(sshRunner *ssh.Runner) (*ResolvFileValues, erro
 func CreateResolvFileOnInstance(sshRunner *ssh.Runner, resolvFileValues ResolvFileValues) error {
        resolvFile, _ := CreateResolvFile(resolvFileValues)
 
+       log.Infof("nameservers: %s", strings.Join(resolvFileValues.GetNameServers(), ","))
        err := sshRunner.CopyDataPrivileged([]byte(resolvFile), "/etc/resolv.conf", 0644)
        if err != nil {
                return fmt.Errorf("Error creating /etc/resolv on instance: %s", err.Error())
diff --git a/pkg/crc/network/types.go b/pkg/crc/network/types.go
index 0d7115a4b..dda58693f 100644
--- a/pkg/crc/network/types.go
+++ b/pkg/crc/network/types.go
@@ -20,6 +20,24 @@ type ResolvFileValues struct {
        NameServers   []NameServer
 }
 
+func (vals *ResolvFileValues) GetSearchDomains() []string {
+       searchDomains := []string{}
+       for _, domain := range vals.SearchDomains {
+               searchDomains = append(searchDomains, domain.Domain)
+       }
+
+       return searchDomains
+}
+
+func (vals *ResolvFileValues) GetNameServers() []string {
+       nameservers := []string{}
+       for _, ns := range vals.NameServers {
+               nameservers = append(nameservers, ns.IPAddress)
+       }
+
+       return nameservers
+}
+
 type Mode string
 
 func (m Mode) String() string {

```

If you want to keep the code you have, can you move it to a helper function? In my opinion, this is a very low-level implementation detail, which is not really interesting to have directly in CreateResolvFileOnInstance body. Your code will segfault if `NameServers` has 0 or 1 elements, this will also need to be fixed if you want to keep it.

praveenkumar replied:
Updated with this api, should prevent segfault.

cfergeau commented on pkg/crc/network/nameservers.go:
```diff
@@ -34,11 +36,35 @@ func GetResolvValuesFromInstance(sshRunner *ssh.Runner) (*ResolvFileValues, erro
+		return nil
 	}
-	return nil
+	if err := sd.Start("ovs-configuration.service"); err != nil {
```
If I followed correctly, `ovs-configuration.service` will be run during cluster startup as a dependency of `systemctl start kubelet`? We start it manually/explicitly because we want to do things after it is running? for kubelet, this won't matter as it will notice `ovs-configuration` is already running, so nothing more to do ?

Have you explored doing something similar to the suggestion in https://access.redhat.com/solutions/6250271 ?
> The file /etc/resolv.conf can be modified using the nmcli utility via a one-shot systemd service that is (executed after `ovs-configuration.service`)

This would avoid the need for manually starting anything systemd related. Maybe we can unconditionally overwrite resolv.conf, and upload the one shot unit file, and that's it, no need for `sd.Status` or `sd.Start`

praveenkumar replied:
>  for kubelet, this won't matter as it will notice ovs-configuration is already running, so nothing more to do ?

Yes, that's right.

>Have you explored doing something similar to the suggestion in https://access.redhat.com/solutions/6250271 ?

Yes, this was also suggested in internal channel ,  If we go the machine config way that means the cluster should able to start and api is accessible but if the resolution doesn't work than even our cluster api would respond and that will not work. We can try to have the unit file as template on crc codebase side which we first update and then start but that also means `sd.Status` and `sd.Start` required.

cfergeau replied:
I am *not* suggesting to use the machine config, I'm only suggesting to use a one-shot systemd service (ie a text file you'd drop in `/etc/systemd/system/` with the right set of `Requires`/`After`/... to have systemd run it automatically after `ovs-configuration.service` but before `kubelet.service`

praveenkumar replied:
you mean to have this unit file template in the crc codebase and we put that in the instance before kubelet service with updated nameserver and search domain?

cfergeau replied:
Yes

praveenkumar replied:
Can we defer it for next sprint? I want this to be in soon so that we can have it tested for coming release. I can create a follow up issue for same.

praveenkumar replied:
https://github.com/crc-org/crc/issues/4152

praveenkumar committed:
network/types: Add helper function to get search domain and nameserver

These helper function return string slice which is useful for next
commit to use it with nmcli command.

cfergeau commented on pkg/crc/network/nameservers.go:
```diff
@@ -34,13 +36,48 @@ func GetResolvValuesFromInstance(sshRunner *ssh.Runner) (*ResolvFileValues, erro
 func CreateResolvFileOnInstance(sshRunner *ssh.Runner, resolvFileValues ResolvFileValues) error {
 	resolvFile, _ := CreateResolvFile(resolvFileValues)
```
This can be moved to the `replaceResolvConfFile` block, or even to the `replaceResolvConf` helper.

praveenkumar replied:
Done

cfergeau commented on pkg/crc/network/nameservers.go:
```diff
@@ -34,13 +36,48 @@ func GetResolvValuesFromInstance(sshRunner *ssh.Runner) (*ResolvFileValues, erro
+	}
+
+	nameservers := strings.Join(resolvFileValues.GetNameServer(), ",")
+	searchDomains := strings.Join(resolvFileValues.GetSearchDomains(), ",")
```
I'd do this in `updateNetworkManagerConfig`

praveenkumar replied:
Done

cfergeau commented on pkg/crc/network/nameservers.go:
```diff
@@ -34,13 +36,48 @@ func GetResolvValuesFromInstance(sshRunner *ssh.Runner) (*ResolvFileValues, erro
 
+func updateNetworkManagerConfig(sshRunner *ssh.Runner, nameservers, searchDomains string) error {
+	// When ovs-configuration service is running we are sure that name of the connection is ovs-if-br-ex
+	_, stderr, err := sshRunner.RunPrivileged("Update resolve.conf file", "nmcli", "con", "modify", "ovs-if-br-ex",
```
`resolve.conf` - typo

Is there a way of programmatically getting the name of the connection to use instead of hardcoding it?

praveenkumar replied:
> Is there a way of programmatically getting the name of the connection to use instead of hardcoding it?

I wish, I tried my search foo but didn't find anything. Let me know if you have some pointer, that would be great.

cfergeau replied:
Might be worth a `nmcli conn show --active ovs-if-br-ex` call first to ensure it exists?

praveenkumar replied:
if it not exist then what would be our fallback? because if it not exist then if will anyway fails when adding the domain and nameserver .

cfergeau replied:
We could output a good error message instead of something possibly hard to understand.

cfergeau commented on pkg/crc/network/nameservers.go:
```diff
@@ -34,13 +36,48 @@ func GetResolvValuesFromInstance(sshRunner *ssh.Runner) (*ResolvFileValues, erro
+	_, stderr, err := sshRunner.RunPrivileged("Update resolve.conf file", "nmcli", "con", "modify", "ovs-if-br-ex",
+		"ipv4.dns", nameservers, "ipv4.dns-search", searchDomains)
+	if err != nil {
+		return fmt.Errorf("failed to update resolve.conf file %s: %v", stderr, err)
```
`resolve.conf`

praveenkumar replied:
Done

cfergeau commented on pkg/crc/network/nameservers.go:
```diff
@@ -32,15 +34,51 @@ func GetResolvValuesFromInstance(sshRunner *ssh.Runner) (*ResolvFileValues, erro
+	// update resolve.conf file
+	if state, err := sd.Status("ovs-configuration.service"); err != nil || state == states.NotFound {
+		if err := replaceResolvConfFile(sshRunner, resolvFileValues); err != nil {
+			return fmt.Errorf("error updating resolve.conf file: %s", err)
```
`resolve.conf` typo

praveenkumar replied:
fixed it.

cfergeau commented on pkg/crc/network/nameservers.go:
```diff
@@ -32,15 +34,51 @@ func GetResolvValuesFromInstance(sshRunner *ssh.Runner) (*ResolvFileValues, erro
+		return fmt.Errorf("failed to update NetworkManager configuration: %s", err)
+	}
+
+	return sd.Restart("NetworkManager.service")
```
Is this required? it's odd if settings done with `nmcli` are not applied immediatly? I'd call this from `updateNetworkManagerConfig`

praveenkumar replied:
Yes it is required.

praveenkumar replied:
updated it to use from `updateNetworkManagerConfig`

cfergeau commented on pkg/crc/network/nameservers.go:
```diff
@@ -32,15 +34,51 @@ func GetResolvValuesFromInstance(sshRunner *ssh.Runner) (*ResolvFileValues, erro
+func updateNetworkManagerConfig(sshRunner *ssh.Runner, resolvFileValues ResolvFileValues) error {
+	nameservers := strings.Join(resolvFileValues.GetNameServer(), ",")
+	searchDomains := strings.Join(resolvFileValues.GetSearchDomains(), ",")
+	// When ovs-configuration service is running we are sure that name of the connection is ovs-if-br-ex
```
I'd replace "we are sure" with "should be" or "is expected to be" :)

praveenkumar replied:
Done

cfergeau commented on pkg/crc/network/nameservers.go:
```diff
@@ -33,7 +33,7 @@ func GetResolvValuesFromInstance(sshRunner *ssh.Runner) (*ResolvFileValues, erro
 }
 
-func CreateResolvFileOnInstance(sshRunner *ssh.Runner, resolvFileValues ResolvFileValues) error {
+func UpdateResolvFileOnInstance(sshRunner *ssh.Runner, resolvFileValues ResolvFileValues) error {
```
One more instance of `resolve.conf` in the commit log.

And maybe split the short log, it's very long
```
nameserver: rename CreateResolvFileOnInstance

It's renamed to UpdateResolvFileOnInstance. The resolv.conf file is always present in the VM, this function ...
```

(I disagree that `cp myresolv.conf /etc/resolv.conf` is an "update", but that's a different matter, not terribly important here ;)
```

praveenkumar replied:
done.

cfergeau commented on pkg/crc/network/nameservers.go:
```diff
@@ -32,15 +34,47 @@ func GetResolvValuesFromInstance(sshRunner *ssh.Runner) (*ResolvFileValues, erro
+	if err != nil {
+		return fmt.Errorf("failed to update resolv.conf file %s: %v", stderr, err)
+	}
+	return sd.Restart("NetworkManager.service")
```
/etc/resolv.conf update can also be achieved with `nmcli conn up ovs-if-br-ex`
